packages/playwright-core/src/utils/traceUtils.ts

```diff
@@ -37,3 +42,43 @@ export function serializeClientSideCallMetadata(metadatas: ClientSideCallMetadat
   }
   return { files: [...fileNames.keys()], stacks };
 }
+
+export async function saveTraceFile(fileName: string, traceEvents: ActionTraceEvent[]) {
+  const lines: string[] = traceEvents.map(e => JSON.stringify(e));
+  const zipFile = new yazl.ZipFile();
+  zipFile.addBuffer(Buffer.from(lines.join('\n')), 'trace.trace');
+
+  const sourceFiles = new Set<string>();
+  for (const event of traceEvents) {
+    for (const frame of event.metadata.stack || [])
+      sourceFiles.add(frame.file);
+  }
+  for (const sourceFile of sourceFiles) {
+    const source = await fs.promises.readFile(sourceFile, 'utf8');
+    zipFile.addBuffer(Buffer.from(source), 'resources/src@' + calculateSha1(sourceFile) + '.txt');
+  }
+  await new Promise(f => {
+    zipFile.end(undefined, () => {
+      zipFile.outputStream.pipe(fs.createWriteStream(fileName)).on('close', f);
+    });
+  });
+}
+
+export function createTraceEventForExpect(apiName: string, expected: any, stack: StackFrame[], wallTime: number): ActionTraceEvent {
+  return {
+    type: 'action',
+    metadata: {
+      id: 'expect@' + wallTime,
```

Comment:
Couldn't there be two expect calls at the same `wallTime`? Do I misunderstand it?